clients/src/main/java/org/apache/kafka/common/internals/SecurityManagerCompatibility.java

```diff
@@ -0,0 +1,355 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.internals;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import javax.security.auth.Subject;
+
+/**
+ * This is a compatibility class to provide dual-support for JREs with and without SecurityManager support.
+ * <p>Users should call {@link #get()} to retrieve a singleton instance, and call instance methods
+ * {@link #doPrivileged(PrivilegedAction)}, {@link #current()}, and {@link #callAs(Subject, Callable)}.
+ * <p>This class's motivation and expected behavior is defined in
+ * <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-1006%3A+Remove+SecurityManager+Support">KIP-1006</a>
+ */
+public interface SecurityManagerCompatibility {
+
+    static SecurityManagerCompatibility get() {
+        return Composite.INSTANCE;
+    }
+
+    <T> T doPrivileged(PrivilegedAction<T> action);
+
+    Subject current();
+
+    <T> T callAs(Subject subject, Callable<T> action) throws CompletionException;
+
+    interface Loader {
+        Class<?> loadClass(String className) throws ClassNotFoundException;
+
+        static Loader forName() {
+            return className -> Class.forName(className, true, Loader.class.getClassLoader());
+        }
+    }
+
+    /**
+     * This class implements reflective access to the deprecated-for-removal methods of AccessController and Subject.
+     * <p>Instantiating this class may fail if any of the required classes or methods are not found.
+     * Method invocations for this class may fail with {@link UnsupportedOperationException} if all methods are found,
+     * but the operation is not permitted to be invoked.
+     * <p>This class is expected to be instantiable in JRE >=8 until the removal finally takes place.
+     */
+    @SuppressWarnings("unchecked")
+    class Legacy implements SecurityManagerCompatibility {
```

Comment:
It'd be nice if the various implementations were all top-level classes in their own files; makes the interface easier to grok the first time through. Maybe all of this logic can be encapsulated in its own Java package with the only public members being the `SecurityManagerCompatibility` interface, its methods, and its `Loader` subinterface?